File: src/main/java/com/intetics/lukyanenko/dao/jdbc/OrderDAOImpl.java
package com.intetics.lukyanenko.dao.jdbc;

import com.intetics.lukyanenko.dao.OrderDAO;
import com.intetics.lukyanenko.models.Order;
import org.springframework.dao.DataAccessException;
import org.springframework.jdbc.core.ResultSetExtractor;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.HashMap;

public class OrderDAOImpl extends CommonDAOImpl<Order> implements OrderDAO
{
  @Override
  protected String getBaseSelectSQL()
  {
    return "select * from order_head";
  }
  
  @Override
  protected Order mapFields(ResultSet resultSet, String fieldNamePrefix)
          throws SQLException
  {
    Order Order = getNewModelInstance();
    Order.setId(resultSet.getInt(fieldNamePrefix + "id"));
    Order.setCreated(resultSet.getDate(fieldNamePrefix + "created"));
    Order.setIsBasket(resultSet.getInt(fieldNamePrefix + "is_basket") != 0);
    Order.setNumber(resultSet.getString(fieldNamePrefix + "order_number"));
    Order.setCustomerId(resultSet.getInt(fieldNamePrefix + "customer_id"));
    return Order;
  }
  
  public void add(final Order order)
  {
    final int id = generateNewID();
    jdbcTemplate.update("insert " +
                        "  into order_head (" +
                        "    id, " +
                        "    order_number, " +
                        "    created, " +
                        "    customer_id, " +
                        "    is_basket " +
                        "  ) values (" +
                        "    :id, " +
                        "    :order_number, " +
                        "    :created, " +
                        "    :customer_id, " +
                        "    :is_basket" +
                        "  )",
                        new HashMap<String, Object>(){{
                          put("id", id);
                          put("order_number", order.getNumber());
                          put("created", order.getCreated());
                          put("customer_id", order.getCustomerId());
                          put("is_basket", order.getIsBasket() ? 1 : 0);
                        }}
                       );
    order.setId(id);
  }
  
  public void update(final Order order)
  {
    jdbcTemplate.update("update order_head " +
                        "  set order_number = :order_number, " +
                        "      customer_id  = :customer_id " +
                        "  where id = :id",
                        new HashMap<String, Object>(){{
                          put("id", order.getId());
                          put("order_number", order.getNumber());
                          put("customer_id", order.getCustomerId());
                        }}
    );
  }
  
  public void delete(final Order order)
  {
    HashMap<String, Object> params = new HashMap<>();
    params.put("id", order.getId());
    jdbcTemplate.update("delete from order_detail where order_head_id = :id", params);
    jdbcTemplate.update("delete from order_head where id = :id", params);
  }
  
  public int generateNewID()
  {
    return jdbcTemplate.queryForObject("select next value for sq_order_head",
                                       (HashMap<String, Object>)null,
                                       Integer.class);
  }
  
  @Override
  public Order findBasket(Integer customerId)
  {
    HashMap<String, Object> params = new HashMap<>(1);
    params.put("customer_id", customerId);
    return jdbcTemplate.query(
      String.format("%s where is_basket = 1 and customer_id = :customer_id", getBaseSelectSQL()),
      params,
      new ResultSetExtractor<Order>()
      {
        public Order extractData(ResultSet resultSet)
          throws SQLException, DataAccessException
        {
          if (resultSet.next())
            return mapFields(resultSet);
          else
            return null;
        }
      }
    );
  }
  
  @Override
  public Order convertBasketToOrder(Order basket)
  {
    HashMap<String, Object> params = new HashMap<>();
    params.put("id", basket.getId());
    jdbcTemplate.update("update order_head set is_basket = 0 where id = :id", params);
    basket.setIsBasket(false);
    return basket;
  }
}

File: src/main/java/com/intetics/lukyanenko/dao/jdbc/OrderDAOImpl.kt
package com.intetics.lukyanenko.dao.jdbc

import com.intetics.lukyanenko.dao.OrderDAO
import com.intetics.lukyanenko.models.Order
import org.springframework.jdbc.core.ResultSetExtractor
import org.springframework.stereotype.Service

import java.sql.ResultSet

@Service
class OrderDAOImpl : CommonDAOImpl<Order>(), OrderDAO {
    private val sqlInsert = """
        insert
          into order_head (
            id,
            order_number,
            created,
            customer_id,
            is_basket
          ) values (
            :id,
            :order_number,
            :created,
            :customer_id,
            :is_basket
          )""".trimIndent()
    private val sqlUpdate = """
        update order_head
          set order_number = :order_number,
              customer_id  = :customer_id
          where id = :id
        """.trimIndent()
    private val sqlDeleteDetails = "delete from order_detail where order_head_id = :id"

    private val sqlDelete = "delete from order_head where id = :id"

    override fun getBaseSelectSQL() = "select * from order_head"

    override fun mapFields(resultSet: ResultSet, fieldNamePrefix: String): Order {
        val Order = newModelInstance()
        Order.id = resultSet.getInt(fieldNamePrefix + "id")
        Order.created = resultSet.getDate(fieldNamePrefix + "created")
        Order.isBasket = resultSet.getInt(fieldNamePrefix + "is_basket") != 0
        Order.number = resultSet.getString(fieldNamePrefix + "order_number")
        Order.customerId = resultSet.getInt(fieldNamePrefix + "customer_id")
        return Order
    }

    override fun add(model: Order) {
        val id = generateNewID()
        jdbcTemplate.update(sqlInsert,
                mapOf("id" to id,
                        "order_number" to model.number,
                        "created" to model.created,
                        "customer_id" to model.customerId,
                        "is_basket" to if (model.isBasket) 1 else 0)
        )
        model.id = id
    }

    override fun update(model: Order) {
        jdbcTemplate.update(sqlUpdate,
                mapOf("id" to model.id, "order_number" to model.number, "customer_id" to model.customerId)
        )
    }

    override fun delete(model: Order) {
        val params = mapOf("id" to model.id)
        jdbcTemplate.update(sqlDeleteDetails, params)
        jdbcTemplate.update(sqlDelete, params)
    }

    override fun generateNewID(): Int {
        return jdbcTemplate.queryForObject("select next value for sq_order_head",
                mapOf<String, Any>(),
                Int::class.java)
    }

    override fun findBasket(customerId: Int): Order? {
        return jdbcTemplate.query(
                String.format("%s where is_basket = 1 and customer_id = :customer_id", getBaseSelectSQL()),
                mapOf("customer_id" to customerId),
                ResultSetExtractor<Order> { resultSet -> if (resultSet.next()) mapFields(resultSet) else null }
        )
    }

    override fun convertBasketToOrder(basket: Order): Order {
        jdbcTemplate.update("update order_head set is_basket = 0 where id = :id", mapOf("id" to basket.id))
        basket.isBasket = false
        return basket
    }
}